File: test/java/com/distributedlife/mahjong/reference/hand/HandCandidateTest.java
package com.distributedlife.mahjong.reference.hand;

import org.junit.Test;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

import static com.distributedlife.mahjong.reference.hand.HandCandidate.times;
import static org.hamcrest.CoreMatchers.is;
import static org.junit.Assert.*;

public class HandCandidateTest {
    @Test
    public void shouldBeValidIfItHas14TilesRequired() {
        ArrayList<String> availableTiles = new ArrayList<String>();
        for(int i = 0; i < 15; i++) {
            availableTiles.add("a");
        }
        HandCandidate handCandidate = new HandCandidate("derp", availableTiles);

        handCandidate.add("a", times(13));
        assertFalse(handCandidate.isValid());

        handCandidate.add("a");
        assertTrue(handCandidate.isValid());

        handCandidate.add("a");
        assertFalse(handCandidate.isValid());
    }

    @Test
    public void canAddShouldBeTrueWhenAvailableTilesContainsThatQuantityOfTile() {
        ArrayList<String> availableTiles = new ArrayList<String>();
        availableTiles.add("a");

        HandCandidate handCandidate = new HandCandidate("derp", availableTiles);
        assertTrue(handCandidate.canAdd("a"));
        assertFalse(handCandidate.canAdd("a", times(2)));
        assertFalse(handCandidate.canAdd("b"));
    }

    @Test
    public void addShouldMoveQuantityOfTilesFromAvailableToRequired() {
        ArrayList<String> availableTiles = new ArrayList<String>();
        availableTiles.add("a");

        HandCandidate handCandidate = new HandCandidate("derp", availableTiles);
        handCandidate.add("a");

        assertThat(handCandidate.getRequiredTiles().size(), is(1));
        assertThat(handCandidate.getRequiredTiles().get(0), is("a"));
        assertFalse(handCandidate.canAdd("a"));
    }

    @Test
    public void shouldForkAllValues() {
        List<String> availableTiles = new ArrayList<String>();
        availableTiles.add("1 Bamboo");
        availableTiles.add("2 Bamboo");
        HandCandidate original = new HandCandidate("hello", availableTiles);
        original.setPrimarySuit("Bamboo");
        original.setSecondSuit("Spot");
        original.setThirdSuit("Crack");
        original.add("1 Bamboo");

        assertThat(original.fork().getName(), is("hello"));
        assertThat(original.fork().getPrimarySuit(), is("Bamboo"));
        assertThat(original.fork().getSecondSuit(), is("Spot"));
        assertThat(original.fork().getThirdSuit(), is("Crack"));
        assertThat(original.fork().getRequiredTiles(), is(Arrays.asList("1 Bamboo")));
        assertThat(original.fork().availableTiles, is(Arrays.asList("2 Bamboo")));
    }
}

File: test/hand-candidate.test.js
/**
 * Tests for HandCandidate
 */
import { describe, it } from 'node:test';
import assert from 'node:assert';
import { HandCandidate, times } from '../lib/hand/hand-candidate.js';

describe('HandCandidate', () => {
  it('should create a hand candidate', () => {
    const tiles = ['1 Bamboo', '2 Bamboo', '3 Bamboo'];
    const candidate = new HandCandidate('Test', tiles);
    
    assert.strictEqual(candidate.getName(), 'Test');
    assert.strictEqual(candidate.getRequiredTiles().length, 0);
  });

  it('should check if tile can be added', () => {
    const tiles = ['1 Bamboo', '2 Bamboo', '3 Bamboo'];
    const candidate = new HandCandidate('Test', tiles);
    
    assert.ok(candidate.canAdd('1 Bamboo'));
    assert.ok(!candidate.canAdd('5 Bamboo'));
  });

  it('should add tiles correctly', () => {
    const tiles = ['1 Bamboo', '1 Bamboo', '1 Bamboo', '2 Bamboo'];
    const candidate = new HandCandidate('Test', tiles);
    
    candidate.add('1 Bamboo', times(3));
    assert.strictEqual(candidate.getRequiredTiles().length, 3);
    assert.ok(!candidate.canAdd('1 Bamboo'));
  });

  it('should fork correctly', () => {
    const tiles = ['1 Bamboo', '2 Bamboo'];
    const candidate = new HandCandidate('Test', tiles);
    candidate.setPrimarySuit('Bamboo');
    candidate.add('1 Bamboo');
    
    const fork = candidate.fork();
    assert.strictEqual(fork.getName(), candidate.getName());
    assert.strictEqual(fork.getPrimarySuit(), candidate.getPrimarySuit());
    assert.strictEqual(fork.getRequiredTiles().length, 1);
  });

  it('should validate hand size', () => {
    const tiles = Array(20).fill('1 Bamboo');
    const candidate = new HandCandidate('Test', tiles);
    
    assert.ok(!candidate.isValid());
    
    for (let i = 0; i < 14; i++) {
      candidate.add('1 Bamboo');
    }
    
    assert.ok(candidate.isValid());
  });
});
